social_core/backends/nfdi.py

```diff
@@ -0,0 +1,212 @@
+"""
+Backend for OpenID Connect OPs of the German NFDI initiative
+https://doc.nfdi-aai.de/
+
+This is conceptually based on the egi_checkin backend
+"""
+
+from social_core.backends.open_id_connect import OpenIdConnectAuth
+
+NFDI_ENDPOINTS = {
+    # AcademicID
+    "xcs": "https://keycloak.sso.gwdg.de/auth/realms/academiccloud",
+    "textplus": "https://keycloak.sso.gwdg.de/auth/realms/academiccloud",
+    # didmos
+    "mardi": "https://auth.didmos.nfdi-aai.de",
+    "objects": "https://auth.didmos.nfdi-aai.de",
+    "culture": "https://auth.aai.nfdi4culture.de",
+    # regapp
+    "cat": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    "chem": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    "datascience": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    "energy": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    "ing": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    "matWerk": "https://regapp.nfdi-aai.de/oidc/realms/nfdi",
+    # unity
+    "daphne": "https://login.helmholtz.de/oauth2",
+    "fairmat": "https://login.helmholtz.de/oauth2",
+    "immuno": "https://login.helmholtz.de/oauth2",
+    "punch": "https://login.helmholtz.de/punch-oauth2",
+    "helmholtz": "https://login.helmholtz.de/oauth2",
+    # infraproxy
+    "infraproxy-staging": "https://infraproxy-staging.nfdi-aai.dfn.de",
+    "infraproxy": "https://infraproxy.nfdi-aai.dfn.de",
+    # eduid
+    "eduid": "https://proxy.edu-id.dfn.de",
+    "eduid-staging": "",
+    # other
+    # DataPLANT
+    # GHGA
+    # NFDI4Biodiversity
+    # unassigned
+    # BERD@NFDI
+    # FAIRagro
+    # KonsortSWD
+    # NFDI4BIOIMAGE
+    # NFDI4Earth
+    # NFDI4Health
+    # NFDI4Memory
+    # NFDI4Microbiota
+}
+
+
+class NFDIOpenIdConnect(OpenIdConnectAuth):
+    name = "helmholtz"
+    OIDC_ENDPOINT = "https://login.helmholtz.de/oauth2"
+    # In order to get any scopes, you have to register your service with
+    # the OP
+    DEFAULT_SCOPE = [
+        "openid",
+        "profile",
+        "email",
+        "voperson_id",
+        "eduperson_entitlement",
+        # "entitlement",
+        "eduperson_scoped_affiliation",
+        "voperson_external_affiliation",
+        "eduperson_assurance",
+        "orcid",
+        # "offline_access",
+    ]
+    # This is the list of entitlements that are allowed to login into the
+    # service. A user with any of these will be allowed. If empty, all
+    # users will be allowed
+    ALLOWED_ENTITLEMENTS = []
+
+    def get_user_details(self, response):
```

Comment:
Can you please add tests for this? You can mock it using `user_data_url` and `user_data_body` attributes on the test class.